### PR TITLE
[Added Feature] Undo Last Entry Feature

### PR DESCRIPTION
**Pull request by dozysushi**
I am pushing the code for the feature which is undo last entry, this feature is responsible for deleting the most recent dataPoint under the selected metric. I do many different test in main class, the funtionality work well and I also doing  a unit basically compared the difference from the metrics has been deleted last entry and the premade metric. I wIll finish UI and more unit tests later.

### CODE REVIEW

**Review by Bensonwu02 — CHANGES_REQUESTED** [review]
You can delete the overloaded constructor for the response model because I don't think you are using it?

I would make a custom exception class to handle the case where the metric contains 0 datapoints. Let me know if you're unsure about how to do this.

File: gradlew.bat [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+:fail
+rem Set variable GRADLE_EXIT_CONSOLE if you need the _script_ return code instead of
+rem the _cmd.exe /c_ return code!
+if  not "" == "%GRADLE_EXIT_CONSOLE%" exit 1
+exit /b 1
+
+:mainEnd
+if "%OS%"=="Windows_NT" endlocal
+
+:omega
```
Comment (Wizard603) [inline gradlew.bat:89]: Can you remove the .idea and gradle related files from you commits.

File: src/main/java/UseCases/EntryUndo.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            }catch (Exception e) {
+                return presenter.prepareFailView("Metric can't undo - " + metricName
+                        + " contains fewer than 1 data point");}
+            DataPoint deletedData = deletedMetric.get(deletedMetric.size() - 1);
+            this.metricStorage.removeDataPoint(metricName);
+            double value = deletedData.getValue();
+            LocalDateTime date = deletedData.getDate();
+            EntryUndoResponseModel responseModel = new EntryUndoResponseModel(value, date, metricName);
+            return presenter.prepareSuccessView(responseModel);
+            } catch (Exception e) {
```
Comment (Wizard603) [inline src/main/java/UseCases/EntryUndo.java:37]: Maybe replace with a custom Exception or use a existing more specific one

File: src/main/java/UseCases/EntryUndo.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    public EntryUndo(MetricStorage metricStorage, EntryUndoOutputBoundary presenter){
+        this.metricStorage = metricStorage;
+        this.presenter = presenter;
+    }
+
+    public EntryUndoResponseModel deleteDatapoint(EntryUndoRequestModel requestModel){
+        String metricName = requestModel.getMetricName();
+        try{
+            ArrayList<DataPoint> deletedMetric = this.metricStorage.getDataPointList(metricName);
```
Comment (Wizard603) [inline src/main/java/UseCases/EntryUndo.java:25]: note getDataPointList can return null

Comment (Wizard603) [reply src/main/java/UseCases/EntryUndo.java:25]: consider adding a null pointer exception or dealing with the empty list situation in some other way. Likely they should not be able to undo anything in this case

**Review by Wizard603 — CHANGES_REQUESTED** [review]
Summary:
- Remove the idea and gradle project files
- Narrow Exception handling to more specific cases
- Consider what happens when null is returned from getDataPointList

also you should clean up the unnecessary new lines for better readability

**Review by tylerkj1209 — APPROVED** [review]

**Comment by dozysushi** [comment]
> You can delete the overloaded constructor for the response model because I don't think you are using it?
> 
> I would make a custom exception class to handle the case where the metric contains 0 datapoints. Let me know if you're unsure about how to do this.

I deleted unused method in response model, will add a exception class later.

File: gradlew.bat (hunk 1 above)
Comment (dozysushi) [reply gradlew.bat:89]: > Summary:
> 
> * Remove the idea and gradle project files
> * Narrow Exception handling to more specific cases
> * Consider what happens when null is returned from getDataPointList
> 
> also you should clean up the unnecessary new lines for better readability

I deleted the idea and gradle, working on exception class, right now I use try to fix the problem

File: src/main/java/UseCases/EntryUndo.java (hunk 3 above)
Comment (dozysushi) [reply src/main/java/UseCases/EntryUndo.java:25]: working on exception class

**Review by Wizard603 — APPROVED** [review]

**Review by Bensonwu02 — APPROVED** [review]

**Comment by dozysushi** [comment]
I think everything is working right now. I fix all the problems from comment.

**Pull request merged**